### PR TITLE
Update to work with the new device registration method.

### PR DESCRIPTION
**Pull request by Page-**

### CODE REVIEW

**Comment by jviotti** [comment]
The Travis CI builds are complaining about `karma` not existing. Since this application no longer uses this, we should remove it from `.travis.yml`: https://github.com/resin-io-modules/resin-register-device/blob/master/.travis.yml#L13

File: tests/register.spec.coffee [hunk 1]
```diff
@@ -1,147 +1,84 @@
...
 					apiKey: 'asdf'
-				, (error, device) =>
-					expect(error).to.equal(@error)
-					expect(device).to.not.exist
+				, (error, deviceInfo) ->
+					expect(error).to.be.an('error').that.has.a.property('message', 'Unauthorized')
+					expect(deviceInfo).to.not.exist
 					done()
 
 			it 'should return a rejected promise if no callback', ->
```
Comment (jviotti) [inline tests/register.spec.coffee]: I believe the test title not longer makes sense.

File: tests/register.spec.coffee [hunk 2]
```diff
@@ -1,147 +1,84 @@
...
-						uuid: 'asdf'
-						apiKey: 'asdf'
+						uuid: 'zxcv'
+						apiKey: 'qwerty'
 					done()
 
-			it 'should return a promise that resolves the device if no callback', ->
-				promise = register.register @pineInstance,
-					userId: 199
+			it 'should return a promise that resolves to the device info if no callback', ->
```
Comment (jviotti) [inline tests/register.spec.coffee]: The `no callback` part should probably be removed as well.

**Comment by jviotti** [comment]
This module also has a SauceLabs badge on the README, that now that karma has been removed, it should probably be removed as well.

**Comment by jviotti** [comment]
The README documentation should also be updated to reflect the new API (and remove `generateUUID`).

**Comment by Page-** [comment]
Ah, I didn't update the README as I thought the docs were automatic from the in-line comment docs, fixed

File: tests/register.spec.coffee (hunk 1 above)
Comment (Page-) [inline tests/register.spec.coffee]: What do you suggest instead?

File: tests/register.spec.coffee (hunk 2 above)
Comment (Page-) [inline tests/register.spec.coffee]: Done

**Comment by Page-** [comment]
And I've removed the karma/saucelabs stuff fully as well

**Comment by jviotti** [comment]
Yeah, this module doesn't have the jsdoc markdown generation feature (I've discovered it after it was created :) ).

LGTM :+1:

File: README.md [hunk 3]
```diff
@@ -22,53 +21,36 @@ $ npm install --save resin-register-device
...
 - `Number userId`: The user id.
 - `Number applicationId`: The application id.
 - `String deviceType`: The device type slug.
+- `String apiEndpoint`: The API endpoint.
 - `String apiKey`: The API key.
-- `[String uuid]`: The uuid. Notice this property is optional, and it will be generated if absent.
-- `[String apiPrefix]`: The API prefix. Notice this property is optional, and it will use the one from the passed pine instance if absent.
 
-The `callback` gets called with two arguments: `(error, device)`, where `device` is an object containing two properties: an `id` and the `uuid`.
+The `callback` gets called with two arguments: `(error, deviceInfo)`, where `deviceInf` is an object containing three properties: the `id`, `uuid`, and `api_key` for the device that was just registered.
```
Comment (emirotin) [inline README.md]: typo, `deviceInf`

File: lib/register.coffee [hunk 4]
```diff
@@ -55,42 +25,39 @@ exports.generateUUID = (callback) ->
...
 # @param {Object} options - options
 # @param {Number} options.userId - user id
 # @param {Number} options.applicationId - application id
 # @param {String} options.deviceType - device type
+# @param {String} options.apiEndpoint - api endpoint
 # @param {String} options.apiKey - api key
-# @param {String} [options.uuid] - device uuid
-# @param {String} [options.apiPrefix] - api prefix
-# @param {Function} callback - callback (error, device)
+# @param {Function} [callback] - callback (error, device)
```
Comment (emirotin) [inline lib/register.coffee]: makes sense to change device to deviceInfo?

Comment (Page-) [inline lib/register.coffee]: Fixed, thanks

File: README.md (hunk 3 above)
Comment (Page-) [inline README.md]: Fixed, thanks

**Comment by pcarranzav** [comment]
Just in case anyone stumbles into this: please do not merge.
We need the uuid to be generated by this module as it is now.

**Comment by Page-** [comment]
Yeah, only I should be merging, and I need to make changes to this whole PR set (there's around 5-6 PRs related to this work)

**Comment by Page-** [comment]
Ok, I've updated this so it should be good now, I still need to update the dependent PRs and test these changes within those, but it should be good to rereview

File: lib/register.coffee [hunk 5]
```diff
@@ -15,37 +15,38 @@ limitations under the License.
 ###
 
 Promise = require('bluebird')
+postAsync = Promise.promisify(require('request').post, multiArgs: true)
 _ = require('lodash')
 
+randomHexChar = ->
+	return _.random(15).toString(16)
```
Comment (pcarranzav) [inline lib/register.coffee]: We need this to be cryptographically secure, and it's not: [lodash uses Math.random](https://github.com/lodash/lodash/blob/4.15.0/lodash.js#L1488) which is [not cryptographically secure](http://stackoverflow.com/a/5651854/2549019)

Comment (pcarranzav) [inline lib/register.coffee]: [Something like his](https://github.com/resin-io/resin-supervisor/blob/master/src/lib/random-hex-string.coffee) should work instead.

File: lib/register.coffee [hunk 6]
```diff
@@ -15,37 +15,38 @@ limitations under the License.
...
+# deviceRegister.generateUniqueKey (error, randomKey) ->
+#  	throw error if error?
+# 	# randomKey is a randomly generated key that can be used as either a uuid or an api key
+# 	console.log(randomKey)
 ###
-exports.generateUUID = (callback) ->
-
-	# I'd be nice if the UUID matched the output of a SHA-256 function,
-	# but although the length limit of the CN attribute in a X.509
+exports.generateUniqueKey = (callback) ->
```
Comment (pcarranzav) [inline lib/register.coffee]: Can we keep generateUUID as an alias for backwards-compatibility?

File: lib/register.coffee (hunk 5 above)
Comment (Page-) [inline lib/register.coffee]: This was using `Math.random` before, check https://github.com/resin-io-modules/resin-register-device/pull/22/files/6dcf9f3a78e1544b2cec84ee792282b166746a01#diff-d23d2f1f8dd77d2219c945046bbaf969L46 - all I did was to make use of lodash functions to reduce the amount of code required to do the same thing  
As for why, I remember a big convo with @petrosagg involved so he can probably elaborate

File: lib/register.coffee (hunk 6 above)
Comment (Page-) [inline lib/register.coffee]: I changed it from async to sync so it wouldn't be backwards compatible even if the old name remained

File: lib/register.coffee (hunk 5 above)
Comment (pcarranzav) [inline lib/register.coffee]: Right, but before it was used only for  uuids and not apikeys, so the need for being crypto. secure wasn't as pressing... now it is.

Comment (emirotin) [inline lib/register.coffee]: I agree here, the crypto-secure generator should be used for the API keys.

File: lib/register.coffee (hunk 6 above)
Comment (pcarranzav) [inline lib/register.coffee]: if we need the crypto secure generator as per the comment above, this will probably go back to being async

Comment (petrosagg) [inline lib/register.coffee]: Can we use node-randomstring? it does the right things. https://github.com/klughammer/node-randomstring

Comment (Page-) [inline lib/register.coffee]: Done

File: README.md [hunk 7]
```diff
@@ -22,53 +21,51 @@ $ npm install --save resin-register-device
...
 - `String deviceType`: The device type slug.
-- `String apiKey`: The API key.
-- `[String uuid]`: The uuid. Notice this property is optional, and it will be generated if absent.
-- `[String apiPrefix]`: The API prefix. Notice this property is optional, and it will use the one from the passed pine instance if absent.
+- `String deviceApiKey`: The API key to create for the newly registered device.
+- `String provisioningApiKey`: The provisioning API Key.
+- `String apiEndpoint`: The API endpoint.
 
-The `callback` gets called with two arguments: `(error, device)`, where `device` is an object containing two properties: an `id` and the `uuid`.
+The `callback` gets called with two arguments: `(error, deviceInfo)`, where `deviceInfo` is an object containing one properties: the `id` for the device that was just registered.
```
Comment (pcarranzav) [inline README.md]: Detail: `one property`

**Comment by pcarranzav** [comment]
LGTM

File: README.md (hunk 7 above)
Comment (Page-) [inline README.md]: Fixed

File: lib/register.coffee (hunk 6 above)
Comment (petrosagg) [inline lib/register.coffee]: `callback` is unused

File: lib/register.coffee [hunk 8]
```diff
@@ -55,42 +56,48 @@ exports.generateUUID = (callback) ->
...
-				registered_at: Math.floor(Date.now() / 1000)
-			customOptions:
-				apikey: options.apiKey
+exports.register = Promise.method (options, callback) ->
+	for opt in [ 'userId', 'applicationId', 'uuid', 'deviceType', 'provisioningApiKey', 'apiEndpoint']
+		if !options[opt]?
+			throw new Error("Options must contain a '#{opt}' entry.")
+	postAsync
+		method: 'POST'
+		url: "#{options.apiEndpoint}/device/register?apikey=#{options.provisioningApiKey}"
```
Comment (petrosagg) [inline lib/register.coffee]: Can we pass the apikey as a header?

Comment (Page-) [inline lib/register.coffee]: Not atm, I believe @lekkas is working on that

File: lib/register.coffee (hunk 6 above)
Comment (Page-) [inline lib/register.coffee]: Removed

File: lib/register.coffee (hunk 8 above)
Comment (lekkas) [inline lib/register.coffee]: Yeah, this is in my queue (there's a spec in HQ)

**Pull request merged**